[LLVM][TableGen] Support type casts of nodes with multiple results

stephenchouca opened the pull request:
Currently, type casts can only be used to pattern match for intrinsics with a single overloaded return value. For instance:
```
def int_foo : Intrinsic<[llvm_anyint_ty], []>;
def : Pat<(i32 (int_foo)), ...>;
```

This patch extends type casts to support matching intrinsics with multiple overloaded return values. As an example, the following defines a pattern that matches only if the overloaded intrinsic call returns an `i16` for the first result and an `i32` for the second result:
```
def int_bar : Intrinsic<[llvm_anyint_ty, llvm_anyint_ty], []>;
def : Pat<([i16, i32] (int_bar)), ...>;
```

topperc commented on llvm/include/llvm/Target/TargetSelectionDAG.td:
```diff
@@ -2070,3 +2070,18 @@ class ComplexPattern<ValueType ty, int numops, string fn,
+// This can be used to pattern match for intrinsics with multiple overloaded
+// return values. For example:
+//
+// def int_foo : Irtrinsic<[llvm_anyint_ty, llvm_anyint_ty], []>;
```
Intrinsic*

stephenchouca replied:
Thanks for catching these typos! Fixed.

topperc commented on llvm/include/llvm/Target/TargetSelectionDAG.td:
```diff
@@ -2070,3 +2070,18 @@ class ComplexPattern<ValueType ty, int numops, string fn,
+//
+
+class ValueTypeList<list<ValueType> vts> {
+  list<ValueType> VTs = vts ;
```
extra space before semicolon

stephenchouca replied:
Fixed.

arsenm commented on llvm/utils/TableGen/Common/CodeGenDAGPatterns.cpp:
```diff
@@ -2903,23 +2903,46 @@ TreePatternNodePtr TreePattern::ParseTreePattern(Init *TheInit,
+    return New;
+  }
+
+  if (Operator->isSubClassOf("ValueTypeList")) {
```
Could you just directly consume list<ValueType> without this wrapper class?

stephenchouca replied:
Changed to just directly consume a list of types.

arsenm commented on llvm/test/TableGen/dag-isel-valuetypelist.td:
```diff
@@ -0,0 +1,35 @@
+// RUN: llvm-tblgen -gen-dag-isel -I %p/../../include %s | FileCheck %s
+
```
Can you also handle the globalisel emitter

topperc replied:
Both emitters share CodeGenDAGPatterns.cpp right? So it should just need a test?

arsenm replied:
Yes, probably

stephenchouca replied:
Added a test for this. As suggested, this didn't require any change to GlobalISel emitter.

arsenm commented on llvm/utils/TableGen/Common/CodeGenDAGPatterns.cpp:
```diff
@@ -2896,6 +2896,35 @@ TreePatternNodePtr TreePattern::ParseTreePattern(Init *TheInit,
   }
+
+  auto ParseCastOperand = [this](DagInit *Dag, StringRef OpName) {
+    if (Dag->getNumArgs() != 1)
+      error("Type cast only takes one operand!");
+
+    if (!OpName.empty())
+      error("Type cast should not have a name!");
+
```
Can you add tests for these error conditions?

stephenchouca replied:
Done.

arsenm commented on llvm/utils/TableGen/Common/CodeGenDAGPatterns.cpp:
```diff
@@ -2896,6 +2896,35 @@ TreePatternNodePtr TreePattern::ParseTreePattern(Init *TheInit,
+    TreePatternNodePtr New = ParseCastOperand(Dag, OpName);
+
+    unsigned NumTypes = New->getNumTypes();
+    if (NumTypes != LI->size())
+      error("Invalid number of type casts!");
```
This too

stephenchouca replied:
Done.

arsenm commented on llvm/test/TableGen/multiple-type-casts-patfrags.td:
```diff
@@ -0,0 +1,38 @@
+// GISEL: GIM_RootCheckType, /*Op*/0, /*Type*/GILLT_s32
+// GISEL: GIM_RootCheckType, /*Op*/1, /*Type*/GILLT_s16
+// GISEL: GIR_BuildRootMI, /*Opcode*/GIMT_Encode2(::INSTR_FOO_I32_I16)
+def : Pat<([i32, i16] (int_foo)), ([i32, i16] (INSTR_FOO_I32_I16))>;
```
Add a test with an empty list?

stephenchouca replied:
I think this should probably be considered invalid as well, since a cast with an empty list would essentially just be a no-op. Added a check and negative test for this too.

arsenm commented on llvm/utils/TableGen/Common/CodeGenDAGPatterns.cpp:
```diff
@@ -2886,6 +2886,35 @@ TreePatternNodePtr TreePattern::ParseTreePattern(Init *TheInit,
+    TreePatternNodePtr New = ParseCastOperand(Dag, OpName);
+
+    size_t NumTypes = New->getNumTypes();
+    if (LI->size() == 0 || LI->size() != NumTypes)
```
->empty if it supports it

stephenchouca replied:
Done.

stephenchouca committed:
[LLVM][TableGen] Support type casts of nodes with multiple results

Currently, type casts can only be used to pattern match for intrinsics with a single overloaded return value. For instance:
```
def int_foo : Intrinsic<[llvm_anyint_ty], []>;
def : Pat<(i32 (int_foo)), ...>;
```

This patch extends type casts to support matching intrinsics with multiple overloaded return values. As an example, the following defines a pattern that matches only if the overloaded intrinsic call returns an `i16` for the first result and an `i32` for the second result:
```
def int_bar : Intrinsic<[llvm_anyint_ty, llvm_anyint_ty], []>;
def : Pat<([i16, i32] (int_bar)), ...>;
```

stephenchouca commented:
It'd be great if someone can merge this pull request for me; I don't have write access. Thanks!
